From 924cbb487bd1d03f3408977696d936228ff52ec9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Benno=20F=C3=BCnfst=C3=BCck?= <benno.fuenfstueck@gmail.com>
Date: Thu, 16 Jun 2016 19:13:04 +0200
Subject: [PATCH] Fix icon update bugs

Fixes the following bugs:

* When the handler for g_timeout_add_seconds returns FALSE, the timer is stopped, so the icon
  will no longer be updated. To fix this bug, make sure we always return TRUE from the handler.
* When we have a playback switch and a MUTE/UNMUTE command is received, we should immediatly update
  the icon afterwards. For that to work, we also have to update vol.mute before calling the function
  to update the icon.
* When the volume is zero, we should still show the icon for low volume (the icon should be shown
  for vol >= 0, not just for vol > 0)
---
 gvolicon.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/gvolicon.c b/gvolicon.c
index 3bf2c8f..6995d45 100644
--- a/gvolicon.c
+++ b/gvolicon.c
@@ -143,7 +143,7 @@ tray_icon_set_icon(GtkStatusIcon *icon) {
 
 	if(!vol.has_playback_switch && vol.realvol == 0)
 		gtk_status_icon_set_from_icon_name(icon, (configuration.icon_type == STANDARD) ? ICON_MUTE : ICON_MUTE_SYMBOLIC);
-	else if(vol.realvol > 0 && vol.realvol <= 33)
+	else if(vol.realvol >= 0 && vol.realvol <= 33)
 		gtk_status_icon_set_from_icon_name(icon, (configuration.icon_type == STANDARD) ? ICON_LOW : ICON_LOW_SYMBOLIC);
 	else if(vol.realvol > 33 && vol.realvol <= 66)
 		gtk_status_icon_set_from_icon_name(icon, (configuration.icon_type == STANDARD) ? ICON_MEDIUM : ICON_MEDIUM_SYMBOLIC);
@@ -192,9 +192,10 @@ tray_icon_on_click(GtkStatusIcon *icon) {
 	snd_mixer_handle_events(handle);
 	elem = snd_mixer_find_selem(handle, vol_info);
 
-	if(vol.has_playback_switch)
+	if(vol.has_playback_switch) {
 		snd_mixer_selem_set_playback_switch_all(elem, !vol.mute);
-	else {
+		vol.mute = !vol.mute;
+	} else {
 		if(vol.realvol != 0) {
 			vol.oldvol = vol.realvol;
 			vol.realvol = 0;
@@ -207,7 +208,7 @@ tray_icon_on_click(GtkStatusIcon *icon) {
 	tray_icon_set_tooltip(icon);
 }
 
-static void
+static gboolean
 tray_icon_check_for_update(GtkStatusIcon *icon)  {
 	snd_mixer_elem_t *elem;
 	long min; //not used, but segfault if NULL in call below
@@ -233,6 +234,7 @@ tray_icon_check_for_update(GtkStatusIcon *icon)  {
 		tray_icon_set_icon(icon);
 		tray_icon_set_tooltip(icon);
 	}
+	return TRUE;
 }
 
 int
-- 
2.8.3

